src/Likvido.ApiAuth.Server/ApiKeyAuthMiddleware.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Likvido.ApiAuth.Server
+{
+    public class ApiKeyAuthMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger<ApiKeyAuthMiddleware> _logger;
+
+        public ApiKeyAuthMiddleware(RequestDelegate next, ILogger<ApiKeyAuthMiddleware> logger)
+        {
+            _next = next;
+            _logger = logger;
+        }
+
+        //TODO: better options approach
+        public async Task Invoke(HttpContext context, IApiKeyAuthOptions options)
+        {
+            if (context.Request.Headers.TryGetValue(options.HeaderName, out var apiKey))
+            {
+                var userName = options.GetUserName(apiKey);
+                if (string.IsNullOrWhiteSpace(userName))
+                {
+                    _logger.ApiKeyAuthorizationFailureOccurred(
+                        "Invalid Api Key.",
+                        new Dictionary<string, object> { ["HeaderName"] = options.HeaderName });
+                    await WriteUnauthorized(context).ConfigureAwait(false);
+                    return;
+                }
+
+                var identity = new ClaimsIdentity(
+                    new[]
+                    {
+                        new Claim(ClaimTypes.Name, userName)
+                    },
+                    options.HeaderName);
+
+                context.User = new ClaimsPrincipal(identity);
+
+                await _next(context).ConfigureAwait(false);
+            }
+            else
+            {
+                _logger.ApiKeyAuthorizationFailureOccurred(
+                    "Api Key not provided.",
+                    new Dictionary<string, object> { ["HeaderName"] = options.HeaderName });
+                await WriteUnauthorized(context).ConfigureAwait(false);
+            }
+        }
+
+        private static Task WriteUnauthorized(HttpContext context)
```

Comment:
What is the reason for returning a task here, and calling it as async?

Reply:
I've been writing error details to a response initially. Then I decided to stop doing this. A consumer can use `StatusCodePages` middleware to return properly shaped error responses. 
But you are right no need in return Task now